In [23]:
import pandas as pd
import numpy as np

# 24036首歌，361346个LEs
train_data = pd.read_csv('./Train_Test_Splits/Context/train.csv')
test_data = pd.read_csv('./Train_Test_Splits/Context/test.csv')
df = pd.concat([train_data, test_data], axis=0)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 361346 entries, 0 to 104333
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   user_id           361346 non-null  int64  
 1   track_id          361346 non-null  object 
 2   hashtag           361346 non-null  object 
 3   created_at        361346 non-null  object 
 4   score             361346 non-null  float64
 5   lang              361346 non-null  object 
 6   tweet_lang        361346 non-null  object 
 7   time_zone         328718 non-null  object 
 8   instrumentalness  361333 non-null  float64
 9   liveness          361333 non-null  float64
 10  speechiness       361332 non-null  float64
 11  danceability      361332 non-null  float64
 12  valence           361330 non-null  float64
 13  loudness          361346 non-null  float64
 14  tempo             361346 non-null  float64
 15  acousticness      361333 non-null  float64
 16  energy            36

In [24]:
spotify_df = pd.read_csv('./data/spotify_dataset.csv')

spotify_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114000 entries, 0 to 113999
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Unnamed: 0        114000 non-null  int64  
 1   track_id          114000 non-null  object 
 2   artists           113999 non-null  object 
 3   album_name        113999 non-null  object 
 4   track_name        113999 non-null  object 
 5   popularity        114000 non-null  int64  
 6   duration_ms       114000 non-null  int64  
 7   explicit          114000 non-null  bool   
 8   danceability      114000 non-null  float64
 9   energy            114000 non-null  float64
 10  key               114000 non-null  int64  
 11  loudness          114000 non-null  float64
 12  mode              114000 non-null  int64  
 13  speechiness       114000 non-null  float64
 14  acousticness      114000 non-null  float64
 15  instrumentalness  114000 non-null  float64
 16  liveness          11

In [25]:
spo_drop = spotify_df.drop_duplicates(subset='track_id')
spo_drop.shape

(89741, 21)

In [26]:
spo_drop.drop('Unnamed: 0', axis = 1, inplace = True)
spo_drop.head(10)

/home/s1751118/.local/lib/python3.10/site-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


track_id                               artists  \
0  5SuOikwiRyPMVoIQDJUgSV                           Gen Hoshino   
1  4qPNDBW1i3p13qLCt0Ki3A                          Ben Woodward   
2  1iJBSr7s7jYXzM8EGcbK5b                Ingrid Michaelson;ZAYN   
3  6lfxq3CG4xtTiEg7opyCyx                          Kina Grannis   
4  5vjLSffimiIP26QG5WcN2K                      Chord Overstreet   
5  01MVOl9KtVTNfFiBU9I7dc                          Tyrone Wells   
6  6Vc5wAMmXdKIAM7WUoEb7N  A Great Big World;Christina Aguilera   
7  1EzrEOXmMH3G43AXT1y7pA                            Jason Mraz   
8  0IktbUcnAGrvD03AWnz3Q8             Jason Mraz;Colbie Caillat   
9  7k9GuJYLp2AzqokyEdwEw2                        Ross Copperman   

                                          album_name  \
0                                             Comedy   
1                                   Ghost (Acoustic)   
2                                     To Begin Again   
3  Crazy Rich Asians (Original Motion Picture Sou...   
4                                            Hold On   
5                               Days I Will Remember   
6                        Is There Anybody Out There?   
7                We Sing. We Dance. We Steal Things.   
8                We Sing. We Dance. We Steal Things.   
9                                             Hunger   

                   track_name  popularity  duration_ms  explicit  \
0                      Comedy          73       230666     False   
1            Ghost - Acoustic          55       149610     False   
2              To Begin Again          57       210826     False   
3  Can't Help Falling In Love          71       201933     False   
4                     Hold On          82       198853     False   
5        Days I Will Remember          58       214240     False   
6               Say Something          74       229400     False   
7                   I'm Yours          80       242946     False   
8                       Lucky          74       189613     False   
9                      Hunger          56       205594     False   

   danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.676  0.4610    1    -6.746     0       0.1430        0.0322   
1         0.420  0.1660    1   -17.235     1       0.0763        0.9240   
2         0.438  0.3590    0    -9.734     1       0.0557        0.2100   
3         0.266  0.0596    0   -18.515     1       0.0363        0.9050   
4         0.618  0.4430    2    -9.681     1       0.0526        0.4690   
5         0.688  0.4810    6    -8.807     1       0.1050        0.2890   
6         0.407  0.1470    2    -8.822     1       0.0355        0.8570   
7         0.703  0.4440   11    -9.331     1       0.0417        0.5590   
8         0.625  0.4140    0    -8.700     1       0.0369        0.2940   
9         0.442  0.6320    1    -6.770     1       0.0295        0.4260   

   instrumentalness  liveness  valence    tempo  time_signature track_genre  
0          0.000001    0.3580   0.7150   87.917               4    acoustic  
1          0.000006    0.1010   0.2670   77.489               4    acoustic  
2          0.000000    0.1170   0.1200   76.332               4    acoustic  
3          0.000071    0.1320   0.1430  181.740               3    acoustic  
4          0.000000    0.0829   0.1670  119.949               4    acoustic  
5          0.000000    0.1890   0.6660   98.017               4    acoustic  
6          0.000003    0.0913   0.0765  141.284               3    acoustic  
7          0.000000    0.0973   0.7120  150.960               4    acoustic  
8          0.000000    0.1510   0.6690  130.088               4    acoustic  
9          0.004190    0.0735   0.1960   78.899               4    acoustic

In [27]:
from sklearn.preprocessing import LabelEncoder

lb = LabelEncoder()
df['track_id'] = lb.fit_transform(df['track_id'])
df_indices = df['track_id'].value_counts()
df_indices

18192    2782
356      2188
19602    2177
9724     2102
22636    1129
         ... 
20083       1
7801        1
3707        1
15997       1
5757        1
Name: track_id, Length: 26911, dtype: int64

In [28]:
# 根据数量信息对 spotify_df 进行采样
sampled_df = spo_drop.iloc[np.repeat(df_indices.index, df_indices.values)].copy()
sampled_df['track_id'].value_counts()

0vwh0gJI3ovQwuMdSl1C8S    2782
4I18rXWixQBWc8JhTqx6Xn    2188
3G1o1dkJIyY38yfiVDkwNP    2177
0lIAAEd3YoxWdEgYEdlrdE    2102
0MYqimbl02aogFBWyx75bV    1129
                          ... 
6Dh6nTzrGg1IsEL6K8PwSP       1
68ZOBzR0AffpQM6AgSA5n4       1
0LZgf4LDn9hHNTtNgKhC0u       1
1jkh0HKmZnBELVtNcWQ3eZ       1
08BdwDsjgDp9N3ALRaKuYJ       1
Name: track_id, Length: 26911, dtype: int64

In [29]:
df['track_id'] = lb.inverse_transform(df['track_id'])
df['track_id'].value_counts()

acca7456f99b4699871d3fe249e5fe11    2782
02f5cf0eb2798bc9a1d3fc3c2c83267b    2188
b9fa88d18f7dda845789155d69177df0    2177
5cbf25f0f194b8710a1ac9844a832d24    2102
d6b1124fdd64c4b1afa59f967397111d    1129
                                    ... 
39009176dd6180db557c81b889d36462       1
f3947da379fac6a3f7bc9ee2c96f8b1f       1
d23cf849930e2e7e0d3ce7c0234065d7       1
63f5e6b7aa5bf19eedf514686f7bb371       1
2bff6c31eb08dae31b62cc4a29c8aa0a       1
Name: track_id, Length: 26911, dtype: int64

In [30]:
series_count_df = df['track_id'].value_counts()
series_count_spotify = sampled_df['track_id'].value_counts()
columns = ['track_id', 'instrumentalness', 'liveness', 'speechiness', 'danceability', 'valence', 'loudness', 'tempo', 'acousticness', 'energy', 'mode', 'key']

In [31]:
# 使用 loc 方法替换 df 中符合条件的行
# df.loc[df['track_id'] == 'acca7456f99b4699871d3fe249e5fe11', ['track_id', 'tempo', 'liveness']] = df_new.iloc[0]
df.reset_index(drop=True, inplace=True)
sampled_df.reset_index(drop=True, inplace=True)
df, sampled_df

(           user_id                          track_id               hashtag  \
 0        823214346  b129f3e0a0bdc1e61c57ce994f9d2932                kiss92   
 1        823214346  64b30f6ac1acecb7cef9dc890e1c3ea7                kiss92   
 2        823214346  cc1a46ee0446538ecf6b65db01c30cd8                kiss92   
 3        823214346  dbbd8525c1bc48cc791bfe618b74bac6                kiss92   
 4        823214346  dbdbb866939c1ab1dfd0d5cfac740eaa                kiss92   
 ...            ...                               ...                   ...   
 361341  1549216658  fbfd4b6ff80084a38f596257089009fc          ripjoecocker   
 361342    33679225  ca6e67e4f5840c389c426f8e11a99dec               hotchip   
 361343   186008229  c4a0c3feae47b6e6a09141483782302f              confused   
 361344   223620983  2fd9f021f0a9336ef8c2585740491d21                bummer   
 361345    26274755  6adc880dc402b81f77b8c7c5b436c303  criminallyunderrated   
 
                  created_at  score lang tweet_lan

In [32]:
mapping = dict(zip(series_count_df.index, series_count_spotify.index))
mapping

{'acca7456f99b4699871d3fe249e5fe11': '0vwh0gJI3ovQwuMdSl1C8S',
 '02f5cf0eb2798bc9a1d3fc3c2c83267b': '4I18rXWixQBWc8JhTqx6Xn',
 'b9fa88d18f7dda845789155d69177df0': '3G1o1dkJIyY38yfiVDkwNP',
 '5cbf25f0f194b8710a1ac9844a832d24': '0lIAAEd3YoxWdEgYEdlrdE',
 'd6b1124fdd64c4b1afa59f967397111d': '0MYqimbl02aogFBWyx75bV',
 '3d02f9fcad37e6bb227682761039498c': '6xVbrVyIuGVgybBHrhN12E',
 '462b1fc2bc97a2c8de40946cba29fd21': '6w23wsDSQWZbdFAXic8krw',
 'b2cb6927d9c7ea77e34a36466c388b1f': '7h3rVSwqhTiFiysLcVlR4S',
 'd67b89478389301e93a416f16331ed6e': '2O64HZaJMurTa9F1jMo8l3',
 '902785553dfe83c414136a31c892a416': '2VZtXhjdZE6etfcXIKt7FB',
 'f28716acc5ba2d609b9300717394d01b': '1IbDJI6psd1UgsoASFxBAQ',
 'e9e56df7298bd09fc2bf4782d111f4a6': '1tAoZwDf76POzOIkeCyAsn',
 '1d954bf6800232e2b42c7ecc1e96d940': '0d28khcov6AiegSCpG5TuT',
 'ba012b0edbc9ef4ace33ecd16ce39b5e': '4VFre5iu3eUunkIpsx7h6W',
 '1827a1398750b0210f5f2394eb56fe47': '3NRVfxtC8mYrPoHjKahcng',
 'a6e7de1836bebd7937ebf25395979ee9': '3iReY6sbqHlvFtd3g

In [33]:
df['track_id'] = df['track_id'].map(mapping)
df_merge = df.merge(spo_drop, on='track_id', how='left')
df_merge

user_id                track_id               hashtag  \
0        823214346  3Px454rzMHtIRc9R04QSlB                kiss92   
1        823214346  5eaD4UEbmcXZe0buNHOM64                kiss92   
2        823214346  75VOiyc0bT0VkFFQuKakgE                kiss92   
3        823214346  0Dy1Q2uUKII4asOHvexBn9                kiss92   
4        823214346  777VRhgQZ1L2c0RJnLlNI9                kiss92   
...            ...                     ...                   ...   
361341  1549216658  3RFL9hE5cBocBsjD5XLJLQ          ripjoecocker   
361342    33679225  2NmGfvtRSYdxUOeJDtf8VN               hotchip   
361343   186008229  73vBHepDFkAXufkyhR9r0O              confused   
361344   223620983  6hVkpZkfQV82W6e31Xye37                bummer   
361345    26274755  5ZzjNL4J31lqU853Un7T3f  criminallyunderrated   

                 created_at  score lang tweet_lang  \
0       2014-01-01 05:59:38   0.80   en         en   
1       2014-01-01 06:06:15   0.80   en         en   
2       2014-01-01 06:11:07   0.80   en         en   
3       2014-01-01 06:19:43   0.80   en         en   
4       2014-01-01 06:23:47   0.80   en         en   
...                     ...    ...  ...        ...   
361341  2014-12-22 23:31:40   0.15   en         en   
361342  2014-12-23 00:39:26   1.00   en         en   
361343  2014-12-23 03:49:11   0.30   en         en   
361344  2014-12-23 04:50:24   0.30   en         en   
361345  2014-12-23 07:18:15   0.30   en         fr   

                          time_zone  instrumentalness_x  liveness_x  ...  \
0                           Beijing            0.000000      0.0764  ...   
1                           Beijing            0.018300      0.1090  ...   
2                           Beijing            0.000000      0.0931  ...   
3                           Beijing            0.000000      0.0909  ...   
4                           Beijing            0.000005      0.0951  ...   
...                             ...                 ...         ...  ...   
361341                       London            0.000143      0.3200  ...   
361342   Pacific Time (US & Canada)            0.000106      0.0757  ...   
361343                       Alaska            0.001010      0.2330  ...   
361344  Mountain Time (US & Canada)            0.091000      0.2830  ...   
361345                       Harare            0.320000      0.3940  ...   

        loudness_y  mode_y  speechiness_y  acousticness_y  instrumentalness_y  \
0           -7.018       0         0.0802         0.42900            0.321000   
1           -5.646       1         0.0461         0.00245            0.000404   
2           -5.433       1         0.0726         0.04040            0.072600   
3           -8.584       0         0.0292         0.17500            0.844000   
4          -13.974       1         0.0313         0.79000            0.000099   
...            ...     ...            ...             ...                 ...   
361341      -5.074       1         0.0937         0.01010            0.000000   
361342      -9.173       1         0.0320         0.13400            0.000115   
361343      -7.700       1         0.3350         0.36600            0.000000   
361344      -8.593       1         0.0282         0.53400            0.000861   
361345      -3.857       1         0.0428         0.00140            0.000119   

        liveness_y  valence_y  tempo_y  time_signature  track_genre  
0           0.1680     0.9200  126.058               4     afrobeat  
1           0.0838     0.1480  105.929               4  black-metal  
2           0.0697     0.4350  127.806               4    breakbeat  
3           0.0987     0.2780  104.993               4   deep-house  
4           0.1400     0.4010  125.927               4     cantopop  
...            ...        ...      ...             ...          ...  
361341      0.2590     0.7840  150.094               4        anime  
361342      0.1080     0.4040  177.003               4     cantopop  
361343      0.1250     0.6020

In [34]:
df_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 361346 entries, 0 to 361345
Data columns (total 39 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   user_id             361346 non-null  int64  
 1   track_id            361346 non-null  object 
 2   hashtag             361346 non-null  object 
 3   created_at          361346 non-null  object 
 4   score               361346 non-null  float64
 5   lang                361346 non-null  object 
 6   tweet_lang          361346 non-null  object 
 7   time_zone           328718 non-null  object 
 8   instrumentalness_x  361333 non-null  float64
 9   liveness_x          361333 non-null  float64
 10  speechiness_x       361332 non-null  float64
 11  danceability_x      361332 non-null  float64
 12  valence_x           361330 non-null  float64
 13  loudness_x          361346 non-null  float64
 14  tempo_x             361346 non-null  float64
 15  acousticness_x      361333 non-nul

In [35]:
# 假设df是你的DataFrame
filtered_df = df_merge[[col for col in df_merge.columns if not col.endswith('_x')]]
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 361346 entries, 0 to 361345
Data columns (total 28 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   user_id             361346 non-null  int64  
 1   track_id            361346 non-null  object 
 2   hashtag             361346 non-null  object 
 3   created_at          361346 non-null  object 
 4   score               361346 non-null  float64
 5   lang                361346 non-null  object 
 6   tweet_lang          361346 non-null  object 
 7   time_zone           328718 non-null  object 
 8   rating              361346 non-null  int64  
 9   artists             361346 non-null  object 
 10  album_name          361346 non-null  object 
 11  track_name          361346 non-null  object 
 12  popularity          361346 non-null  int64  
 13  duration_ms         361346 non-null  int64  
 14  explicit            361346 non-null  bool   
 15  danceability_y      361346 non-nul

In [36]:
# 假设df是你的DataFrame
filtered_df.columns = [col[:-2] if col.endswith('_y') else col for col in filtered_df.columns]
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 361346 entries, 0 to 361345
Data columns (total 28 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   user_id           361346 non-null  int64  
 1   track_id          361346 non-null  object 
 2   hashtag           361346 non-null  object 
 3   created_at        361346 non-null  object 
 4   score             361346 non-null  float64
 5   lang              361346 non-null  object 
 6   tweet_lang        361346 non-null  object 
 7   time_zone         328718 non-null  object 
 8   rating            361346 non-null  int64  
 9   artists           361346 non-null  object 
 10  album_name        361346 non-null  object 
 11  track_name        361346 non-null  object 
 12  popularity        361346 non-null  int64  
 13  duration_ms       361346 non-null  int64  
 14  explicit          361346 non-null  bool   
 15  danceability      361346 non-null  float64
 16  energy            36

In [37]:
filtered_df = filtered_df[[col for col in filtered_df.columns if col in df.columns]]
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 361346 entries, 0 to 361345
Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   user_id           361346 non-null  int64  
 1   track_id          361346 non-null  object 
 2   hashtag           361346 non-null  object 
 3   created_at        361346 non-null  object 
 4   score             361346 non-null  float64
 5   lang              361346 non-null  object 
 6   tweet_lang        361346 non-null  object 
 7   time_zone         328718 non-null  object 
 8   rating            361346 non-null  int64  
 9   danceability      361346 non-null  float64
 10  energy            361346 non-null  float64
 11  key               361346 non-null  int64  
 12  loudness          361346 non-null  float64
 13  mode              361346 non-null  int64  
 14  speechiness       361346 non-null  float64
 15  acousticness      361346 non-null  float64
 16  instrumentalness  36

In [40]:
filtered_df.head(10)

user_id                track_id hashtag           created_at  score lang  \
0  823214346  3Px454rzMHtIRc9R04QSlB  kiss92  2014-01-01 05:59:38    0.8   en   
1  823214346  5eaD4UEbmcXZe0buNHOM64  kiss92  2014-01-01 06:06:15    0.8   en   
2  823214346  75VOiyc0bT0VkFFQuKakgE  kiss92  2014-01-01 06:11:07    0.8   en   
3  823214346  0Dy1Q2uUKII4asOHvexBn9  kiss92  2014-01-01 06:19:43    0.8   en   
4  823214346  777VRhgQZ1L2c0RJnLlNI9  kiss92  2014-01-01 06:23:47    0.8   en   
5  823214346  6xZMdgMwGxBAJ3O9LfWunY  kiss92  2014-01-01 06:25:48    0.8   en   
6  823214346  0pNeVovbiZHkulpGeOx1Gj  kiss92  2014-01-01 06:42:05    0.8   en   
7  823214346  0hSTmN84rBWsU52Lu6lCj8  kiss92  2014-01-01 06:48:10    0.8   en   
8  823214346  66bK78V5FvJcxdLncAwZ8Q  kiss92  2014-01-01 06:52:15    0.8   en   
9  823214346  3DWNcWoQ2LLqgPYEDaCwkG  kiss92  2014-01-01 06:58:29    0.8   en   

  tweet_lang time_zone  rating  danceability  energy  key  loudness  mode  \
0         en   Beijing       1         0.618   0.831    9    -7.018     0   
1         en   Beijing       1         0.239   0.829    5    -5.646     1   
2         en   Beijing       1         0.627   0.966    7    -5.433     1   
3         en   Beijing       1         0.756   0.462    7    -8.584     0   
4         en   Beijing       1         0.607   0.301    4   -13.974     1   
5         en   Beijing       1         0.926   0.471    2    -9.568     1   
6         en   Beijing       1         0.396   0.338    0   -10.872     1   
7         en   Beijing       1         0.813   0.509    4   -12.016     0   
8         en   Beijing       1         0.172   0.371   11   -10.516     0   
9         en   Beijing       1         0.786   0.721   10    -6.940     1   

   speechiness  acousticness  instrumentalness  liveness  valence    tempo  
0       0.0802      0.429000          0.321000    0.1680    0.920  126.058  
1       0.0461      0.002450          0.000404    0.0838    0.148  105.929  
2       0.0726      0.040400          0.072600    0.0697    0.435  127.806  
3       0.0292      0.175000          0.844000    0.0987    0.278  104.993  
4       0.0313      0.790000          0.000099    0.1400    0.401  125.927  
5       0.1120      0.000432          0.008470    0.1040    0.476  123.979  
6       0.0304      0.198000          0.000003    0.1380    0.377  133.522  
7       0.4630      0.766000          0.000000    0.1040    0.554   94.228  
8       0.0330      0.904000          0.921000    0.1070    0.101   63.998  
9       0.0625      0.235000          0.000870    0.1200    0.891  130.802

In [39]:
filtered_df.to_csv('./data/merge_data.csv', index=False)

Data columns (total 20 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   user_id           361346 non-null  int64  
 1   track_id          361346 non-null  object 
 2   hashtag           361346 non-null  object 
 3   created_at        361346 non-null  object 
 4   score             361346 non-null  float64
 5   lang              361346 non-null  object 
 6   tweet_lang        361346 non-null  object 
 7   time_zone         328718 non-null  object 
 8   instrumentalness  361333 non-null  float64
 9   liveness          361333 non-null  float64
 10  speechiness       361332 non-null  float64
 11  danceability      361332 non-null  float64
 12  valence           361330 non-null  float64
 13  loudness          361346 non-null  float64
 14  tempo             361346 non-null  float64
 15  acousticness      361333 non-null  float64
 16  energy            361333 non-null  float64
 17  mode              361333 non-null  float64
 18  key               361333 non-null  float64
 19  rating            361346 non-null  int64  
dtypes: float64(12), int64(2), object(6)
memory usage: 57.9+ MB